src/book/book.rs

```diff
@@ -0,0 +1,365 @@
+use std::fmt::{self, Display, Formatter};
+use std::path::{Path, PathBuf};
+use std::collections::VecDeque;
+use std::fs::File;
+use std::io::Read;
+
+use super::summary::{parse_summary, Summary, Link, SummaryItem, SectionNumber};
+use errors::*;
+
+
+/// Load a book into memory from its `src/` directory.
+pub fn load_book<P: AsRef<Path>>(src_dir: P) -> Result<Book> {
+    let src_dir = src_dir.as_ref();
+    let summary_md = src_dir.join("SUMMARY.md");
+
+    let mut summary_content = String::new();
+    File::open(summary_md)
+        .chain_err(|| "Couldn't open SUMMARY.md")?
+        .read_to_string(&mut summary_content)?;
+
+    let summary = parse_summary(&summary_content).chain_err(
+        || "Summary parsing failed",
+    )?;
+
+    load_book_from_disk(&summary, src_dir)
+}
+
+
+/// A dumb tree structure representing a book.
+///
+/// For the moment a book is just a collection of `BookItems`.
+#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
+pub struct Book {
+    /// The sections in this book.
+    pub sections: Vec<BookItem>,
+}
+
+impl Book {
+    /// Create an empty book.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Get a depth-first iterator over the items in the book.
+    pub fn iter(&self) -> BookItems {
+        BookItems { items: self.sections.iter().collect() }
+    }
+}
+
+/// Enum representing any type of item which can be added to a book.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum BookItem {
+    /// A nested chapter.
+    Chapter(Chapter),
+    /// A section separator.
+    Separator,
+}
+
+/// The representation of a "chapter", usually mapping to a single file on
+/// disk however it may contain multiple sub-chapters.
+#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
+pub struct Chapter {
+    /// The chapter's name.
+    pub name: String,
+    /// The chapter's contents.
+    pub content: String,
+    /// The chapter's section number, if it has one.
+    pub number: Option<SectionNumber>,
```

Comment:
I wanted to join the old `BookItem::Affix` and `BookItem::Chapter` variants into one thing because they were essentially identical apart from the presence of a section number. 

Do you think I should go back to making a `BookItem` something like this?

```rust
enum BookItem {
  Affix(Chapter),
  NumberedChapter(SectionNumber, Chapter),
  Separator,
}
```

Reply:
I'm not sure, you are right that they are almost identical. The issue here is that we introduce yet another check. First we need to match on the `BookItem` then we need to see if `number` is `Some(_)`.

Reply:
I kind of liked the original distinction for 
```rust
enum BookItem {
  Affix(Chapter),
  NumberedChapter(SectionNumber, Chapter),
  Separator,
}
```
current solution with `Option<SectionNumber>` is a little implicit but I cannot find anything wrong with it.

the variant with separate lists is also nice but enforces that the non numbered chapters should be always pre and post the numbered chapters. I would prefer to leave the decision to the users.

Reply:
>the variant with separate lists is also nice but enforces that the non numbered chapters should be always pre and post the numbered chapters. I would prefer to leave the decision to the users.

That is already a constraint currently. Would you let the user intersperse unnumbered chapters in between numbered chapters? Does numbering restart after that?

Reply:
That's a very good point, it's probably better to adjust the `BookItem` type and the `Book` struct so that the preface, numbered chapters, and appendices are three separate lists. 

Hmm... How would you deal with the types then? From what I can see, you want the ability to have *only* bare chapters (i.e. without chapter numbers) and separators in the appendices and preface, but then *only* numbered chapters and separators in the numbered chapters section. If we use the `BookItem` definition which contains `Affix`, `NumberedChapter`, and `Separator` then in either the numbered section or the outsides you're going to be doing a match where one variant is technically `unreachable!()`. Would we then need two enums? 

It almost feels like in an attempt to avoid the `Option<SectionNumber>` problem we've gone and brought up an even bigger one...

Reply:
> That is already a constraint currently. 

Yeah after acquainting myself with this part of code I see it now.

> Would you let the user intersperse unnumbered chapters in between numbered chapters? Does numbering restart after that?

I would but only if anyone would need such feature. I see no reason to invent artificial use-cases so lets just forget about my previous comment ;)